x/rollapp/keeper/rollapp.go

```diff
@@ -1,13 +1,230 @@
 package keeper
 
 import (
+	"errors"
 	"fmt"
 
+	errorsmod "cosmossdk.io/errors"
 	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
+
 	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
 )
 
+func (k Keeper) RegisterRollapp(ctx sdk.Context, rollapp types.Rollapp) error {
+	if err := rollapp.ValidateBasic(); err != nil {
+		return fmt.Errorf("validate rollapp: %w", err)
+	}
+
+	if err := k.checkIfRollappExists(ctx, rollapp.RollappId, rollapp.Alias); err != nil {
+		return err
+	}
+
+	creator, _ := sdk.AccAddressFromBech32(rollapp.Creator)
+	registrationFee := sdk.NewCoins(k.RegistrationFee(ctx))
+
+	if !registrationFee.IsZero() {
+		if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, creator, types.ModuleName, registrationFee); err != nil {
+			return errors.Join(types.ErrFeePayment, err)
+		}
+
+		if err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, registrationFee); err != nil {
+			return fmt.Errorf("burn coins: %w", err)
+		}
+	}
+
+	k.SetRollapp(ctx, rollapp)
+
+	if err := ctx.EventManager().EmitTypedEvent(&rollapp); err != nil {
+		return fmt.Errorf("emit event: %w", err)
+	}
+
+	return nil
+}
+
+func (k Keeper) UpdateRollapp(ctx sdk.Context, update *types.UpdateRollappInformation) error {
+	if err := update.ValidateBasic(); err != nil {
+		return fmt.Errorf("validate update: %w", err)
+	}
+
+	updated, err := k.canUpdateRollapp(ctx, update)
+	if err != nil {
+		return err
+	}
+
+	k.SetRollapp(ctx, updated)
+
+	// Emit event
+	if err = ctx.EventManager().EmitTypedEvent(&updated); err != nil {
+		return fmt.Errorf("emit event: %w", err)
+	}
+
+	return nil
+}
+
+func (k Keeper) canUpdateRollapp(ctx sdk.Context, update *types.UpdateRollappInformation) (current types.Rollapp, err error) {
+	var found bool
+	current, found = k.GetRollapp(ctx, update.RollappId)
+	if !found {
+		return current, errRollappNotFound
+	}
+
+	if update.Creator != current.Creator {
+		return current, sdkerrors.ErrUnauthorized
+	}
+
+	if current.Frozen {
+		return current, types.ErrRollappFrozen
+	}
+
+	_, hasState := k.GetLatestStateInfoIndex(ctx, update.RollappId)
+	initSequencerBonded := k.sequencerKeeper.IsSequencerBonded(ctx, update.InitialSequencerAddress)
```

Comment:
combine both to single `isMutable`
u can check for violation here, no need to pass it to `canUpdate...` functions.
and u should check for any bonded sequencer

Reply:
if we make it immutable when any bonded, we will not be able to update the immutable fields, because the initial sequencer has to be the first proposer, meaning any sequencer can be bonded even before the initial sequencer, but they cannot be the proposer before the initial sequencer

Reply:
as for `isMutable`, I made errors separate when trying to update when bonded and when there is state so it's clearer